### ISSUE TITLE
Readonly helper incompatible with private class fields.

### ISSUE THREAD

**Issue by b6i6o6** [microsoft/TypeScript#42399]
# Bug Report

### 🔎 Search Terms

private class field readonly helper missing property duck typing

### 🕗 Version & Regression Information

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.2.0-beta#code/MYGwhgzhAEAa0G8BQ1XQA4FcBGICWwGO+wACgE4D260AvNACwBMA3EgL5JLCUB2EAF2hUA+gA8AXNABKAUzAATPiACeAHlgA+OtADy2AFazgAgHQAzcrNkAvWQApesgO5x7ASndtufQcOfiUvD0omIs0AD0EdAAKgAWeDDOlOQA1hCmXKCQMACaiChoWLgERCVkVDT0zGxo0ADE6OR4AG5gArIU1FK8mAC22LLk3nU8-ALkmCYpHgV1dQIJGY3NbR1dVdBMrIXz0GMQlCCypiCUAOb2i4mmK63tnZVeu9CcnD7jwpQiKlJyisp1LltPR9EYTBYrLYHE5XLkPM8Pn5yAFftB8iFvipwlFYkt9pQ+uhwHh+NAwNhKJghH1EhBSedoItZNAAER3NaPaisjCVIYCFSmIA)

### 💻 Code

```ts
class X {
    public publicProp = 42;
}

const ro_x: Readonly<X> = Object.freeze(new X());

const rw_x: X = ro_x; // This works.

class Y {
    public publicProp = 42;
    #privateProp: number;

    constructor() {
        this.#privateProp = 22;
        console.log(this.#privateProp);
    }
}

const ro_y: Readonly<Y> = Object.freeze(new Y());

const rw_y: Y = ro_y; // This complains about missing the "#privateProp" property.
```

### 🙁 Actual behavior

The compiler complains that the readonly type cannot be assigned to the normal type. To be honest, this is something that I could relate to if it was working like that in all cases. For example if the error was thrown for X as well. It would make sense since you expect to get a "normal" X but you get a readonly one instead, things are not likely to go well. But Typescript accepts it, fine. I can accept that as well. But as soon as the type in question has a private class fields then it all falls apart. I just wanted to type the fact that the type properties were readonly, why does the private class field get in the way ?

I tried to type a custom "ReadonlyY" for Y to put back the missing #private property but nope:
```ts
interface ReadonlyY extends Readonly<Y> {
    #privateProp: number; // Private identifiers are not allowed outside class bodies.(18016)
}
```

It's quite funny actually, as the private class fields are invisible to `Readonly`, forbidden in `interface` since that would make it visible, but still somewhat visible to the compiler when comparing a type with its readonly counterpart.

Is there a workaround other than replacing private class fields with standard TS private field ?

### 🙂 Expected behavior

The same behavior for X and Y, regardless of the presence of private class fields.

**Comment by MartinJohns** [comment]
You'll have the same issue with `private` and `protected` properties. Mapped types only operate on properties.

Duplicate of #16441. Used search terms: `readonly private in:title`

**Comment by b6i6o6** [comment]
I agree that it is a duplicate of https://github.com/microsoft/TypeScript/issues/16441. I thought it was related to private class fields, hence I did not look for the right keywords it seems. However, https://github.com/microsoft/TypeScript/issues/16441 was closed with no solution or workaround.

**Comment by MartinJohns** [comment]
There is no solution or workaround. It's a design limitation.

**Comment by b6i6o6** [comment]
While it is a bit sad, I understand that there is no *current* solution or workaround. Is there any plan to overhaul that in the future ? It's rather limiting that private fields be completely incompatible with `Readonly`. Both seem to me to be "basic" features, it is a bit weird to find that they cannot be use together.

Actually, it looks like Typescript would benefit for another way of setting a type as read-only that would not rely on mapped types for the job. I'm just wildly thinking here. Something like this:
```ts
class X {
    public publicProp = 42;
}

const ro_x: readonly X = Object.freeze(new X());

const rw_x: X = ro_x; // This might or might not work, but the behavior would be consistent with the one below.

class Y {
    public publicProp = 42;
    #privateProp: number;

    constructor() {
        this.#privateProp = 22;
        console.log(this.#privateProp);
    }
}

const ro_y: readonly Y = Object.freeze(new Y());

const rw_y: Y = ro_y; // This might or might not work, but the behavior would be consistent with the one above.
```

**Comment by RyanCavanaugh** [comment]
Generally we think of mapped types as corresponding to the runtime operation of doing something like
```ts
function mapObj(value: T, func: (f: T[K]) => Q) {
  const res= { };
  for (let key in value) res[key] = func(res[key]);
  return res;
}
```
In this view, there's no preservation of private or protected properties.

**Comment by b6i6o6** [comment]
I understand that. However, this means that if I want that my readonly class be assignable to its own non-readonly type, then I am forbidden to use any protected or private property. This means I cannot use getter/setter for any of them as I would have nowhere to store the actual value. That's quite restrictive in my opinion.

What is the point of checking the presence of the private properties ? Genuinely asking. It feels strange to me to consider them for some typing operations (type checking) and not consider them for other typing operations (mapped types).

**Comment by RyanCavanaugh** [comment]
> What is the point of checking the presence of the private properties ?

```ts
class A {
    name = "bob";
    private x = Math.random();

    biggerOf(other: A) {
        return this.x > other.x ? this.x : other.x;
    }
}

const p: Readonly<A> = { name: "fred" };
// Unsound: value 'undefined' inhabits binding 'n' of type 'number'
let n: number = p.biggerOf(new A());
```

**Comment by b6i6o6** [comment]
Well, it turns out this feature is useful to me in another case where I want to enforce to use the constructor of the class rather than just having the interface of the class. For the curious:
```ts
// Let's say we have a type (here a class for simplicity) which has special behaviors like getters/setters on some of its members.
class A {
    base: { prop1: number, prop2: string };
    
    constructor(base: { prop1: number, prop2: string }) {
        this.base = { ...base };
        let prop1 = base.prop1;
        Object.defineProperty(base, 'prop1', {
            enumerable: true,
            get() {
                return prop1;
            },
            set(value: number) {
                prop1 = value;
                // Do something in the setter, for example:
                base.prop2 = '' + prop1;
            }
        })
    }
}

const a1: A = new A({ prop1: 28, prop2: 'whatever' });
const a2: A = { base: { prop1: 42, prop2: 'hi' } }; // Oops, I can assign it to the variable ! But the setter won't be called !

// Now if we add a private property to the class... (this could have been done in A but I'm keeping this code shorter)
class B extends A {
    #B!: B;
}

const b1: B = new B({ prop1: 28, prop2: 'whatever' });
const b2: B = { base: { prop1: 42, prop2: 'hi' } }; // Yes, the compiler warns me that something is wrong !
```
